gips/0080-substreams-on-the-network.md

```diff
@@ -0,0 +1,112 @@
+---
+GIP: "0080"
+Title: Integrate Substreams into The Graph Network
+Authors: Alexandre Bourget <alex@streamingfast.io>
+Created: 2024-10-27
+Updated: 2024-10-27
+Stage: Draft
+Discussions-To: *To be added later*
+Category: Protocol Logic, Protocol Interfaces
+Implementations: *To be added later*
+---
+
+*This document tries to express and summarize the discussions that happened in-person at the October 2024 R&D Retreat.*
+
+# Abstract
+
+This GIP proposes integrating Substreams, a streaming-first blockchain indexing technology, into The Graph Network. This integration will enable Substreams to be fully recognized on the network and eventually receive indexing rewards.
+
+# Motivation
+
+Substreams provide significant performance improvements for indexing, particularly for large datasets and complex computations. Integrating Substreams into the network will allow developers to leverage these performance gains while maintaining the decentralization and security benefits of The Graph, as well as encouraging the market to use The Graph as its prime infrastructure provider.
+
+# Detailed Specification
+
+## Indexer Selection Algorithm (ISA)
+
+Given that there's no gateway involved with Substreams (it's direct point-to-point between consumer and provider), the existing Indexer Selection Algorithm (ISA) does not apply.  A round-robin selection of providers will be used in the front-end (https://thegraph.market) to ensure fair distribution of request loads, while still allowing consumers to choose a provider.
+
+## Payments
+
+Payments for Substreams work will be on-chain. The existing `collect` function in the staking contract will be used for payment processing.  It was deemed that no TAP was required to have Substreams collect payments and honor the promises of the protocol.
+
+## Economic Security and Disputes
+
+Economic security will be enforced through slashing and disputes. Indexers providing incorrect data can have their stake slashed.
+
+- **Attestations:** Substreams endpoints will provide signed attestations (optional) to verify data integrity. The command-line tools (`substreams gui --request-attestations` or `substreams run --request-attestations`) will generate and display these attestations.  To accommodate different future attestation methods, and not only an Ethereum key signature, the attestation payload will be prefixed with the protocol used (e.g., `eth:` for Ethereum). Detailed documentation will describe the attestation format and verification process.
```

Comment:
Attestations being optional sounds odd. It makes me think a malicious indexer could provide incorrect data all the time, and only serve correct data when an attestation is provided. Can't attestations be included in payloads by default?

Reply:
Yes perhaps, the server could configure to always have them. I was hesitant to make them always sign off, depending on performance issues.

Also, if someone is running the tech internally, perhaps it doesn't make sense for them to sign stuff.
So perhaps a server serving on the network could have --always-sign-payloads or something?

I'll add some details

Reply:
It makes sense for those who registered their endpoints through the DataEdge contract do run with that parameter all the time. That info should be made accessible via the `/health` or `/info` endpoints mentioned so that at least frontends like thegraph.market could indicate that to consumers (and keep track of those who may at times turn it off). 
But I'd prefer to defer to software defaults and have it enabled, as we should optimize for a scenario where Substreams services are always offered through our network (or at least assume most will do it via The Graph).